src/SDK/MonitoringApplianceEmulator/Trace/LogArchiveTracer.cs

```diff
@@ -0,0 +1,146 @@
+        /// </summary>
+        /// <param name="message">The message to trace</param>
+        public void TraceInformation(string message)
+        {
+            this.Trace(TraceLevel.Info, message);
+        }
+
+        /// <summary>
+        /// Trace <paramref name="message"/> as Error message.
+        /// </summary>
+        /// <param name="message">The message to trace</param>
+        public void TraceError(string message)
+        {
+            this.Trace(TraceLevel.Error, message);
+        }
+
+        /// <summary>
+        /// Trace <paramref name="message"/> as Verbose message.
+        /// </summary>
+        /// <param name="message">The message to trace</param>
+        public void TraceVerbose(string message)
+        {
+            this.Trace(TraceLevel.Verbose, message);
+        }
+
+        /// <summary>
+        /// Trace <paramref name="message"/> as Warning message.
+        /// </summary>
+        /// <param name="message">The message to trace</param>
+        public void TraceWarning(string message)
+        {
+            this.Trace(TraceLevel.Warning, message);
+        }
+
+        /// <summary>
+        /// Adds a custom property, to be included in all traces.
+        /// </summary>
+        /// <param name="name">The property name</param>
+        /// <param name="value">The property value</param>
+        public void AddCustomProperty(string name, string value)
+        {
+        }
+
+        /// <summary>
+        /// Gets the custom properties that are set in this tracer instance.
+        /// </summary>
+        /// <returns>The custom properties, as a read-only dictionary</returns>
+        public IReadOnlyDictionary<string, string> GetCustomProperties()
+        {
+            return new Dictionary<string, string>();
+        }
+
+        #endregion
+
+        #region Implementation of IDisposable
+
+        /// <summary>Performs application-defined tasks associated with freeing, releasing, or resetting unmanaged resources.</summary>
+        public void Dispose()
+        {
+            this.tracerStreamWriter?.Dispose();
+            this.tracerStreamWriter = null;
+
+            this.logFileStream?.Dispose();
+            this.logFileStream = null;
+
+            this.logZipArchive?.Dispose();
+            this.logZipArchive = null;
+        }
+
+        #endregion
+
+        #region Private helper methods
+
+        /// <summary>
+        /// Trace <paramref name="message"/> with the specified level.
+        /// </summary>
+        /// <param name="level">The trace level.</param>
+        /// <param name="message">The message to trace.</param>
+        private void Trace(TraceLevel level, string message)
+        {
+            lock (this.logFileLock)
+            {
+                // Write the trace line to the log file
+                DateTime timestamp = DateTime.UtcNow;
+                this.tracerStreamWriter.WriteLine($"{level}|{timestamp:yyyy-MM-dd HH:mm:ss.fffZ}|{message}");
```

Comment:
>this.tracerStreamWriter.WriteLine($"{level}|{timestamp:yyyy-MM-dd HH:mm:ss.fffZ}|{message}"); [](start = 16, length = 93)

I think that the format of the string should be in the PageableLog object, since that's the object that will read the string and parse it. How about adding a PageableLog.ComposeTraceLine, that will convert from TraceLine to string (reverse of ParseTraceLine)?

Reply:
Moved to TraceLine both the parsing and composing

---
In reply to: [231003594](https://github.com/Azure/SmartDetectorsMonitoringAppliance/pull/13#discussion_r231003594) [](ancestors = 231003594)